ADR 175 Remove assets from Renderer's build. Addressables Integration (#182)
decentraland/adr: content/ADR-175-remove-assets-from-build-local-addressables.md

---
layout: adr
adr: 175
title: Remove assets from Renderer's build. Addressables Integration
date: 2023-01-16
status: Review
type: Standards Track
spdx-license: CC0-1.0
authors:
  - ajimenezDCL
  - mikhail-dcl
---

## Abstract
This document describes an approach on how to make the reference's client lightweight by streaming the currently embedded assets. 

## Problem

> Every nice UI, sound effect, particle or material we add to the renderer increases its size. This directly impact loading times or caching since the client's size will go above the max allowed for it. The idea is to pack textures, audio files and other assets, and stream them in runtime. 

Author: I added a link

## Solution
### Pack the Assets

> Similar to what we are doing with scene assets, we can pack everything into one or multiple (to always be under the cache size limit) asset packs to be consumed by the client. This solution should be part of a different system than the one used for scene asset.

Reviewer: it would be nice to have at least small remark why we can't re-use scene asset approach and these solutions should be different

Author: I think it's about completely different systems: one is served for user-generated content and another is entirely under our control.
I have doubts if and what I should write here. Considering it's a piece from @AjimenezDCL , maybe you have any ideas?

Author: >other stuff
Can you elaborate more? I will put it in the ADR

Reviewer 2: Like the library which handles a ref-counted collection to dispose and release loaded items, built-in tools to read the metadata and metrics files and it's based in promises, which we hate hahaha.

Author: Addressed

<img alt="image" src="https://user-images.githubusercontent.com/118179774/214870799-4fe1b14c-9b17-4485-8958-5ed11eeafdd1.png">

  - the remote section of the profile we are going to ignore
  - the local section refers to building and loading assets from the "Streaming Assets" directory
  - on WebGL "Streaming Assets" are hosted in the same storage where the main binary (wasm) is located, and, thus, are loaded via "Web Request"
  - on Standalone "Streaming Assets" is a directory alongside a main binary, and, thus, are loaded by reading from disk

### Retrocompatibility and Versioning
Currently the renderer is a self-contained experience; and as for "Local" Addressables no dedicated versioning is required:
  - on WebGL "Streaming Assets" are a part of the npm package
  - on Standalone "Streaming Assets" are distributed within the same archive

### Addressables loading
Depending on the specific case, we will use all possible ways to load assets:
  - [Label](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/Labels.html) (to load a list of similar assets)
  - Address (to load an asset by the hardcoded string)

>   - GUID (to load an asset by "Asset Reference")

Reviewer: I think a small example screenshot from Adressables group and Inspector would be beneficial. Just to highlight where 
 Address and Labels are defined.

Reviewer: also, it would be nice to outline our standard of how we define Address - do we keep it as it assigned by default by asset path (I wouldn't recommend it because asset can be moved to a different folder), or we shorten it to more path-independent address

Reviewer: As for GUID reference, we can add just two words that it is "mainly used through Inspector assignment for [SerializedField] AssetReference"

Reviewer: As for GUID would be nice to mention, that the use-case of it is referencing in Inpsector (via public or [sfield] AssetReference)

Reviewer: coming back to the Address topic and looking to it from the another side: I remember it was suggested as a good practice that folder structure of Addressables assets should be equal to the grouping structure inside Addressables groups window... Anyway, we need some standardization here to keep it under control

Author: >As for GUID would be nice to mention, that the use-case of it is referencing in Inpsector (via public or [sfield] AssetReference)

It is mentioned as "(to load an asset by "Asset Reference")" and Asset References themselves are mentioned in a separate section along the link to the Untiy doc

   - We may come up with additional rules for our convenience

Check [Unity guidelines](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/AddressableAssetsDevelopmentCycle.html#organizing-addressable-assets) for additional information.

#### Asset Bundle CRC	
At the moment we don't have any sensitive data that requires integrity check. Moreover, for bigger files it will cause noticeable overhead on loading.
So for the groups we will disable it.

### Resources handling
We are going to exclude Scenes and Resources from Addressables Catalog:
   - We don't plan to migrate "Resources"' API to the corresponding one of "Addressables"
   - We plan to migrate assets currently added to "Resources" to "Addressables" with a proper distribution over time

![image](https://user-images.githubusercontent.com/118179774/214880545-4ecec0c7-6c50-4a57-97dc-946305719ae7.png)

### Addressables Caching
"Addressables" uses right the same "Caching" system that "Asset Bundles" does. It works out of the box for all the supported platforms. It is possible to disable caching per group, but in our case we will not be doing it.

### Asset Reference

> We are going to utilize ["AssetReference"](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/AssetReferences.html) and "AssetLabelReference" instead of "SerializeField" to properly reference Assets that should be a part of an Addressable without creating a strong reference to them.

Reviewer: this is a bit not clear for me. Do you mean that we will go with `AssetReference` via `AssetLabelReference` and not via Inspector injection (which will be forbidden, and probably even forced by the validation tests)?

Author: It's not a replacement for "SerializeField", it serves a different purpose: provide a reference to a separate "Addressable" that is external to the context it is referenced from.

Author: I added 2 use cases

---
layout: adr
adr: 175
title: Remove assets from Renderer's build. Addressables Integration
date: 2023-01-16
status: Review
type: Standards Track
spdx-license: CC0-1.0
authors:
  - ajimenezDCL
  - mikhail-dcl
---

## Abstract
This document describes an approach on how to make the reference's client lightweight by streaming the currently embedded assets. 

## Problem

> Every nice UI, sound effect, particle or material we add to the renderer increases its size. This directly impact loading times or caching since the client's size will go above the max allowed for it (applied to WebGL, check [the PoC](https://github.com/2fd/poc-browser-cache) for details). The idea is to pack textures, audio files and other assets, and stream them in runtime. 

Reviewer 3: ```suggestion
Every nice UI, sound effect, particle, or material we add to the renderer increases its size. This impacts directly loading times or caching since the client's size will go above the max allowed for it (applied to WebGL, check [the PoC](https://github.com/2fd/poc-browser-cache) for details). The idea is to pack textures, audio files, and other assets and stream them in runtime. 
```

## Solution
### Pack the Assets

> Similar to what we are doing with scene assets, we can pack everything into one or multiple (to always be under the cache size limit) asset packs to be consumed by the client. This solution should be part of a different system than the one used for scene asset. Reasons why we can't reuse the system include, but are not limited to:

Reviewer 3: ```suggestion
Similar to what we are doing with scene assets, we can pack everything into one or multiple (to always be under the cache size limit) asset packs to be consumed by the client. This solution should be part of a different system than the one used for scene assets. Reasons why we can't reuse the system include, but are not limited to:
```

>    - The existing system is legacy both from the runtime and distribution sides (it is based on "Asset Bundles"). Some parts require special attention such as refactoring, and some do not, but we are not going to base the new system on the legacy foundation, neither we are going to address the old issues immediately

Reviewer 3: ```suggestion
   - The existing system is legacy both from the runtime and distribution sides (it is based on "Asset Bundles"). Some parts require special attention, such as refactoring, and some do not, but we are not going to ground the new system on the legacy foundation, neither to address the old issues immediately
```

Author: Do you think "ground" is a proper word? I think it has a slightly different connotation, and here we need something closer to "build" or "found"

Reviewer 3: You're right. Lets use something closer to "build" or "found".

   - The system is designed for user-generated content
   - It's designed for a different lifecycle: assets are dynamically released and loaded based on the currently active scenes and other usages. This is redundant for the current purpose, at least for the current iteration

>    - It provides mechanisms for loading and handling unprocessed content whereas the system under consideration is dedicated to compiled assets

Reviewer 3: ```suggestion
   - It provides mechanisms for loading and handling unprocessed content, whereas the system under consideration is dedicated to compiled assets
```

### [Addressables](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/index.html)
The Addressables system provides tools and scripts to organize and package content for an application and an API to load and release assets at runtime.

> It's an additional abstraction layer over the Asset Bundles System but it's not limited to this. "Addressables" gives the flexibility to adjust where to host assets and provides mechanisms for dependency resolution, reference counting, simulation, validation, and diagnostics out of the box.

Reviewer 3: ```suggestion
It's an additional abstraction layer over the Asset Bundles System, but it's not limited to this. "Addressables" gives the flexibility to adjust where to host assets and provides mechanisms for dependency resolution, reference counting, simulation, validation, and diagnostics out of the box.
```

### Strategy

> In the first iteration we are going to use the "Local" mode only:

Reviewer 3: ```suggestion
In the first iteration, we are going to use the "Local" mode only:
```

>   - the default addressables profile is suitable for this purpose, we don't have to introduce our own

Reviewer 3: ```suggestion
  - the default addressables profile is suitable for this purpose, so we don't have to introduce our own
```

<figure>
  <img src="/resources/ADR-175/AddressableProfiles.png" />
</figure>

  - the remote section of the profile we are going to ignore
  - the local section refers to building and loading assets from the "Streaming Assets" directory

>   - on WebGL "Streaming Assets" are hosted in the same storage where the main binary (wasm) is located, and, thus, are loaded via "Web Request". To be able to use the "Local" configuration, the "Streaming Assets" folder path has to be specified in the config section of `loader.ts`. This will give the necessary url needed for Unity to complete the corresponding "Web Request".

Reviewer 3: ```suggestion
  - on WebGL, "Streaming Assets" are hosted in the same storage where the main binary (wasm) is located and, thus, are loaded via "Web Request". To use the "Local" configuration, the "Streaming Assets" folder path has to be specified in the config section of `loader.ts`. This will give the necessary URL needed for Unity to complete the corresponding "Web Request".
```

  - on Standalone "Streaming Assets" is a directory alongside a main binary, and, thus, are loaded by reading from disk

### Retrocompatibility and Versioning

> Currently the renderer is a self-contained experience; and as for "Local" Addressables no dedicated versioning is required:

Reviewer 3: ```suggestion
Currently, the renderer is a self-contained experience; and as for "Local" Addressables, no dedicated versioning is required:
```

>   - on WebGL "Streaming Assets" are a part of the npm package

Reviewer 3: ```suggestion
  - on WebGL, "Streaming Assets" are a part of the npm package
```

>   - on Standalone "Streaming Assets" are distributed within the same archive

Reviewer 3: ```suggestion
  - on Standalone, "Streaming Assets" are distributed within the same archive
```

### Addressables loading
Depending on the specific case, we will use all possible ways to load assets:
  - [Label](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/Labels.html) (to load a list of similar assets)
  - Address (to load an asset by the hardcoded string)
     - By default, the address is a path to the file
<figure>
  <img src="/resources/ADR-175/AddressableAddress.png" />
</figure>

     - It is desirable to shorten the address and adjust it according to the directories and groups structure. The exact rules of the former will be elaborated on in the production pipeline
  - GUID (to load an asset by "Asset Reference")

### [Groups configuration](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/GroupSettings.html)
Group settings determine how the assets in a group are treated in content builds.

<figure>
  <img src="/resources/ADR-175/GroupsConfiguration.png" />
</figure>

We are going to distribute assets in "Addressable Asset Groups" according to the following rules:
   - "Essentials" corresponds to the assets that are mandatory for the application to function (e.g. Skybox) and can't be excluded

>    - Other assets according to the logical designation (e.g. "Fonts") and usage (e.g. "Window XXX Resources") fall into different groups; without them the application can function

Reviewer 3: ```suggestion
   - Other assets according to the logical designation (e.g. "Fonts") and usage (e.g. "Window XXX Resources") fall into different groups; without them, the application can function
```

>    - Assets packed together/separately according to the required loading way: any combination of "Include Addresses in Catalog", "Include GUIDs in Catalog" and "Include Labels in Catalog". Thus, we will reduce the size of Catalog.

Reviewer 3: ```suggestion
   - Assets packed together/separately according to the required loading way: any combination of "Include Addresses in Catalog", "Include GUIDs in Catalog", and "Include Labels in Catalog". Thus, we will reduce the size of the Catalog.
```

   - Bundle Mode "Pack Together by Label" will be used to differentiate between platforms (WebGL, Standalone) within the same logical group to exclude unused assets (and the related overhead) if needed
   - We may come up with additional rules for our convenience

Check [Unity guidelines](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/AddressableAssetsDevelopmentCycle.html#organizing-addressable-assets) for additional information.

#### Asset Bundle CRC	

> At the moment we don't have any sensitive data that requires integrity check. Moreover, for bigger files it will cause noticeable overhead on loading.

Reviewer 3: ```suggestion
At the moment, we don't have any sensitive data that requires an integrity check. Moreover, for bigger files, it will cause noticeable overhead on loading.
```

> So for the groups we will disable it.

Reviewer 3: ```suggestion
So for the groups, we will disable it.
```

### Resources handling
We are going to exclude Scenes and Resources from Addressables Catalog:

>    - We don't plan to migrate "Resources"' API to the corresponding one of "Addressables"

Reviewer 3: ```suggestion
   - We don't plan to migrate the "Resources"' API to the corresponding one of "Addressables"
```

   - We plan to migrate assets currently added to "Resources" to "Addressables" with a proper distribution over time

<figure>
  <img src="/resources/ADR-175/ResourceHandling.png" />
</figure>

### Addressables Caching

> "Addressables" uses right the same "Caching" system that "Asset Bundles" does. It works out of the box for all the supported platforms. It is possible to disable caching per group, but in our case we will not be doing it.

Reviewer 3: ```suggestion
"Addressables" uses right the same "Caching" system that "Asset Bundles" does. It works out of the box for all the supported platforms. It is possible to disable caching per group, but in our case, we will not be doing it.
```

### Asset Reference
We are going to utilize ["AssetReference"](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/AssetReferences.html) and "AssetLabelReference" instead of "SerializeField" to properly reference Assets that should be a part of a separate Addressable without creating a strong reference to them.

We can think of several use cases:
   - Optional references that are loaded on demand in conditional branches, and, thus, should not be present in memory all the time
   - Avoiding the case of referencing the same assets by strong references (that are included in the build) and including them in the Addressable/Asset Bundle in order to prevent duplication

Real use cases and distribution of such assets in groups will be refined and defined according to the feature production pipeline with performance in mind.

### Addressables Disposal

> Currently, we don't have a robust mechanism for cleaning up memory from assets. We are not going to introduce it this time either. Thus, we will keep all the Addressables loaded in memory. Considering it's the "Local" mode only, it's not gonna be a problem.

Reviewer 3: ```suggestion
Currently, we don't have a robust mechanism for cleaning up memory from assets. We are not going to introduce it this time either. Thus, we will keep all the Addressables loaded in memory. Considering it's only the "Local" mode, it's not gonna be a problem.
```

### Simulation

> We will use "Simulate Groups" mode for loading Addressables in the Editor. It will help us detect and investigate problems earlier. We will prefer it over "Asset Database" false mode.

Reviewer 3: ```suggestion
We will use the "Simulate Groups" mode for loading Addressables in the Editor. It will help us detect and investigate problems earlier. We will prefer it over the "Asset Database" false mode.
```

> Simulate Groups mode (BuildScriptVirtualMode) analyzes content for layout and dependencies without creating asset bundles. Assets load from the asset database through the ResourceManager, as if they were loaded through bundles. To see when bundles load or unload during game play, view the asset usage in the Addressables Event Viewer window (Window > Asset Management > Addressables > Event Viewer).

Reviewer 3: ```suggestion
Simulate Groups mode (BuildScriptVirtualMode) analyzes content for layout and dependencies without creating asset bundles. Assets load from the asset database through the ResourceManager, as if they were loaded through bundles. To see when bundles load or unload during gameplay, view the asset usage in the Addressables Event Viewer window (Window > Asset Management > Addressables > Event Viewer).
```

Simulate Groups mode will help us simulate load strategies and tweak our content groups to find the right balance for a production release.

### Addressables Validation

> "Addressables" inherits an issue of assets duplication from the Asset Bundles System. Unity [thoroughly described the problem and the way of its resolution](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/ManagingAssets.html#asset-and-assetbundle-dependencies).

Reviewer 3: ```suggestion
"Addressables" inherits an issue of assets duplication from the Asset Bundles System. Unity [thoroughly described the problem and its resolution](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/ManagingAssets.html#asset-and-assetbundle-dependencies).
```

### [Addressables](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/index.html)
The Addressables system provides tools and scripts to organize and package content for an application and an API to load and release assets at runtime.
It's an additional abstraction layer over the Asset Bundles System but it's not limited to this. "Addressables" gives the flexibility to adjust where to host assets and provides mechanisms for dependency resolution, reference counting, simulation, validation, and diagnostics out of the box.

### Strategy
In the first iteration we are going to use the "Local" mode only:
  - the default addressables profile is suitable for this purpose, we don't have to introduce our own
<img alt="image" src="https://user-images.githubusercontent.com/118179774/214870799-4fe1b14c-9b17-4485-8958-5ed11eeafdd1.png">

  - the remote section of the profile we are going to ignore
  - the local section refers to building and loading assets from the "Streaming Assets" directory

>   - on WebGL "Streaming Assets" are hosted in the same storage where the main binary (wasm) is located, and, thus, are loaded via "Web Request"

Reviewer 4: I think its important to add where and how the StreamingAssets are definied in the config file. I had to modify it to use it in the cjk fonts; otherwise the default did not work. And it can be pretty hidden if you dont know where to look.

https://github.com/decentraland/unity-renderer/blob/4b383bd926143197baccf0a94dd5a501b61b5b74/browser-interface/src/index.ts#L97

Author: Can we add this link along with some explanations of how it works?

Reviewer 4: Done

  - on Standalone "Streaming Assets" is a directory alongside a main binary, and, thus, are loaded by reading from disk

### Retrocompatibility and Versioning
Currently the renderer is a self-contained experience; and as for "Local" Addressables no dedicated versioning is required:
  - on WebGL "Streaming Assets" are a part of the npm package
  - on Standalone "Streaming Assets" are distributed within the same archive

### Addressables loading

> Depending on the specific case, we will use all possible ways to load assets:

Reviewer 4: Does it make sense to explain how Im using the `IAddressableResourceProvider` service to retrieve addressables? Is so, let me know and I ll add it here.

Author: I think this ADR is higher-level and is not bound to the implementation of our services

  - [Label](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/Labels.html) (to load a list of similar assets)
  - Address (to load an asset by the hardcoded string)
  - GUID (to load an asset by "Asset Reference")

> ### [Groups configuration](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/GroupSettings.html)

Reviewer 5: In the settings, you can define CRC check along with caching. In general, for local assets, a CRC check is not very needed, unless we have sensitive information. In that case, we should also handle it gracefully if we detect that. However, for our use case, I would consider disabling it for local assets, as bigger files can give quite a significant performance drop during loading.

Author: What does it mean "gracefully" in this case?
In general I see no use and will specify that

Author: I added a small section with clarifications

![image](https://user-images.githubusercontent.com/118179774/214880545-4ecec0c7-6c50-4a57-97dc-946305719ae7.png)

### Addressables Caching
"Addressables" uses right the same "Caching" system that "Asset Bundles" does. It works out of the box for all the supported platforms. It is possible to disable caching per group, but in our case we will not be doing it.

### Asset Reference
We are going to utilize ["AssetReference"](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/AssetReferences.html) and "AssetLabelReference" instead of "SerializeField" to properly reference Assets that should be a part of an Addressable without creating a strong reference to them.
Proper use cases and distribution of such assets in groups will be defined according to the feature production pipeline with performance in mind.

### Addressables Disposal
Currently, we don't have a robust mechanism for cleaning up memory from assets. We are not going to introduce it this time either. Thus, we will keep all the Addressables loaded in memory. Considering it's the "Local" mode only, it's not gonna be a problem.

### Simulation

> We will use "Simualte Groups" mode for loading Addressables in the Editor. It will help us detect and investigate problems earlier. We will prefer it over "Asset Database" false mode.

Reviewer 4: Typo, Simualte

Simulate Groups mode will help us simulate load strategies and tweak our content groups to find the right balance for a production release.

### Addressables Validation
"Addressables" inherits an issue of assets duplication from the Asset Bundles System. Unity [thoroughly described the problem and the way of its resolution](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/ManagingAssets.html#asset-and-assetbundle-dependencies).
"Addressables" provides the ["Analyze Tool"](https://docs.unity3d.com/Packages/com.unity.addressables@1.21/manual/AnalyzeTool.html) and several built-in presets to help detect and/or mitigate such issues.

We have to launch the analysis and fix the issues before pushing Addressables to prevent duplicates and unnecessary implicit references. 

<img alt="image" src="https://user-images.githubusercontent.com/118179774/215037501-46764cb9-6059-45c9-b37d-e2c4571356e1.png">

### CI/CD

> "Streaming Assets" directory should be delivered as an artifact within the same package. 

Reviewer 4: Do you think its useful to reference that this is done here?

https://github.com/decentraland/unity-renderer/blob/4b383bd926143197baccf0a94dd5a501b61b5b74/browser-interface/scripts/build.ts#L50

Author: Can we explain it in words so it's understandable for anybody who takes a look on it for the first time?

Reviewer 4: Done

Author: I like the way you put it, thank you @dalkia
